stdlib/path.gr

```diff
@@ -0,0 +1,729 @@
+    (Abs(_), Rel(_)) | (Abs(_), Rel(_)) => Err(Incompatible(DifferentBases)),
+    _ => Result.map(toPath, relativeTo(pathInfo1, pathInfo2)),
+  }
+}
+
+let rec segsAncestry = (baseSegs, pathSegs) =>
+  match ((baseSegs, pathSegs)) {
+    ([], []) => Self,
+    ([], _) => Descendant,
+    (_, []) => Ancestor,
+    ([first1, ..._], [first2, ..._]) when first1 != first2 => NoLineage,
+    ([_, ...rest1], [_, ...rest2]) => segsAncestry(rest1, rest2),
+  }
+
+// should be used on paths with same absolute/relativeness
+let ancestry = (base: PathInfo, path: PathInfo) => {
+  let (b1, _, s1) = base
+  let (b2, _, s2) = path
+  match ((b1, b2)) {
+    (Abs(d1), Abs(d2)) when d1 != d2 => Err(DifferentRoots),
+    _ => Ok(segsAncestry(List.reverse(s1), List.reverse(s2))),
+  }
+}
+
+/**
+ * Determines the relative ancestry betwen two paths.
+ * 
+ * @param base: The first path to consider
+ * @param path: The second path to consider
+ * @returns `Ok(ancestryStatus)` with the relative ancestry between the paths if they are compatible, or `Err(err)` if they are incompatible
+ * 
+ * @example ancestry(fromString("/usr"), fromString("/usr/bin/bash")) == Ok(Ancestor)
+ * @example ancestry(fromString("/Users/me"), fromString("/Users")) == Ok(Descendant)
+ * @example ancestry(fromString("/usr"), fromString("/etc")) == Ok(Neither)
+ * @example ancestry(fromString("C:/dir1"), fromString("/dir2")) == Err(DifferentRoots)
+ * 
+ * @since v0.5.5
+ */
+export let ancestry = (path1: Path, path2: Path) => {
+  let pathInfo1 = pathInfo(path1)
+  let (base1, _, _) = pathInfo1
+  let pathInfo2 = pathInfo(path2)
+  let (base2, _, _) = pathInfo2
+  match ((base1, base2)) {
+    (Rel(_), Abs(_)) | (Abs(_), Rel(_)) => Err(DifferentBases),
+    _ => ancestry(pathInfo1, pathInfo2),
+  }
+}
+
+let parent = (path: PathInfo) =>
+  match (path) {
+    (base, _, [_, ...rest]) => (base, Dir, rest),
+    (Rel(upDirs), _, []) => (Rel(upDirs + 1), Dir, []),
+    (Abs(_) as base, _, []) => (base, Dir, []),
+  }: PathInfo
+
+/**
+ * Retrieves the path corresponding to the parent directory of the given path.
+ * 
+ * @param path: The path to inspect
+ * @returns A path corresponding to the parent directory of the given path
+ * 
+ * @example parent(fromString("./dir/inner")) == fromString("./dir/")
+ * @example parent(fromString("/")) == fromString("/")
+ * 
+ * @since v0.5.5
+ */
+export let parent = (path: Path) => {
+  toPath(parent(pathInfo(path)))
+}
+
+let basename = (path: PathInfo) =>
+  match (path) {
+    (_, _, [name, ..._]) => Some(name),
+    _ => None,
+  }
+
+/**
+ * Retrieves the basename (named final segment) of a path.
+ * 
+ * @param path: The path to inspect
+ * @returns `Some(path)` containing the basename of the path, or `None` if the path does not have one
+ * 
+ * @example basename(fromString("./dir/file.txt")) == Some("file.txt")
+ * @example basename(fromString(".."))) == None
+ * 
+ * @since v0.5.5
+ */
+export let basename = (path: Path) => {
```

Comment:
Do we want basename to return `Option`? I think it's the only function in this module that returns `Option` now, is that correct?

Reply:
Yeah I believe so; Another option is to make it return empty string instead for paths with no clear name or also possibly `".."` or `"."`. Which approach do you think would be best?

Reply:
Why do we want option instead of Result?

Reply:
Hmm, I typically think of `Result` as being for operations that could fail as a result of malformed input or other conditions, whereas `Option` for valid operations that just don't have the value you are asking for. This case seems more like the latter to me, although maybe you have a alternate perspective?

Reply:
Let's get some feedback from the rest of the team 😄

Reply:
I think option makes sense! In the case that there's no basename, it's not really that it failed, just that there isn't one.